### ISSUE TITLE
Test failure with null when multiple nested with blocks

### ISSUE THREAD

**Issue by cylon** [spockframework/spock#778]
# Issue description
When using deeply nested with() blocks the test fails with a null value

    Condition not satisfied:

    with(it[0]) { with(it[0]) { it == 'end' } }
    |    | |
    null | [end]
         [[end]]

# How to reproduce
https://github.com/cylon/spockwithtest

# Additional Environment information
gradle 4.2

## Java/JDK
java version "1.8.0_144"
Java(TM) SE Runtime Environment (build 1.8.0_144-b01)
Java HotSpot(TM) 64-Bit Server VM (build 25.144-b01, mixed mode)

## Groovy version
2.4.10

### Gradle
------------------------------------------------------------
Gradle 4.2
------------------------------------------------------------

Build time:   2017-09-20 14:48:23 UTC
Revision:     5ba503cc17748671c83ce35d7da1cffd6e24dfbd

Groovy:       2.4.11
Ant:          Apache Ant(TM) version 1.9.6 compiled on June 29 2015
JVM:          1.8.0_144 (Oracle Corporation 25.144-b01)
OS:           Windows 10 10.0 amd64

## Operating System
Windows

### Gradle/Grails
    compile 'org.codehaus.groovy:groovy-all:2.4.10'
    testCompile 'org.spockframework:spock-core:1.1-groovy-2.4'

**Comment by leonard84** [comment]
According to your example project this works with spock-1.0 and not anymore with spock-1.1 correct? I'm guessing that it was broken by 21002e672f5bd552e3eb6f981b723821cdf97d81 however this needs more investigation. @angry-cellophane could you look at it?

**Comment by cylon** [comment]
Yes, that is correct in 1.0 it worked and in 1.1 it doesn't.

**Comment by leonard84** [comment]
@angry-cellophane the problem you described was this line `SpockRuntime.verifyMethodCondition(this.getThisObject(), "contains", [1])`
> Because the method is invoked on this but not the closure it leads to an error (Method not found).

Did you try to change it to
`SpockRuntime.verifyMethodCondition(this, "contains", [1])`
instead?

**Comment by leonard84** [comment]
`git bisect` verified that 21002e672f5bd552e3eb6f981b723821cdf97d81 is the culprit. When I reverted this commit it fixed the problem for master.

So we need a better way then to use the identity closure workaround.

**Comment by angry-cellophane** [comment]
Finally had enough time to look into the issue.
```
SpockRuntime.verifyMethodCondition(this, "contains", [1])
```
the code above doesn't do what we want: to invoke the contains method on the closure. If we leave this groovy will use a reference to the spec at this place.

The bug was caused by the PR #606 
The problem with the with expressions is that they are part of spec, with is a method of spock.lang.Specification and should always be invoked on the spec. The PR changed that behavior.  
I've raised another PR to exclude the with expressions from being invoked on the closure and leave them being invoked on the spec.
